okhttp-tests/src/test/java/okhttp3/DuplexTest.java

```diff
@@ -0,0 +1,183 @@
+        .addHeader("h2", "v2")
+        .setBody(new DuplexResponseBody() {
+          @Override public Header.Listener onRequest(RecordedRequest request,
+              BufferedSource requestBodySource, HttpSink responseBodySink) throws IOException {
+
+            latchParty.step(1);
+            responseBodySink.sink().writeUtf8("ok");
+            responseBodySink.sink().flush();
+            responseBodySink.headers(Headers.of("b", "banada"));
+
+            latchParty.step(3);
+            responseBodySink.sink().writeUtf8("http");
+            responseBodySink.sink().flush();
+
+            latchParty.step(5);
+            responseBodySink.headers(Headers.of("c", "cookie"));
+            responseBodySink.sink().close();
+
+            return null;
+          }
+        }));
+    enableProtocol(Protocol.HTTP_2);
+
+    Call call = client.newCall(new Request.Builder()
+        .url(server.url("/"))
+        .duplex("POST")
+        .build());
+
+    Response response = call.execute();
+
+    RecordingHeadersListener headersListener = new RecordingHeadersListener();
+    response.headersListener(headersListener);
+
+    assertEquals(Headers.of("h1", "v1", "h2", "v2"), response.headers());
+
+    latchParty.step(2);
+    BufferedSource source = response.body().source();
+    assertEquals("ok", source.readUtf8(2));
+
+    latchParty.step(4);
+    assertEquals("http", source.readUtf8(4));
+    assertEquals(Headers.of("b", "banada"), headersListener.takeFirst());
+
+    latchParty.step(6);
+    assertTrue(source.exhausted());
+    assertEquals(Collections.singletonList(Headers.of("c", "cookie")), headersListener.takeAll());
+  }
+
+  @Test public void serverReadsHeadersDataHeadersData() throws IOException {
+    final RecordingHeadersListener serverHeadersListener = new RecordingHeadersListener();
+    final AtomicReference<BufferedSource> requestBodySourceRef = new AtomicReference<>();
+
+    server.enqueue(new MockResponse()
+        .clearHeaders()
+        .addHeader("h1", "v1")
+        .addHeader("h2", "v2")
+        .setBody(new DuplexResponseBody() {
+          @Override public Header.Listener onRequest(RecordedRequest request,
+              BufferedSource requestBodySource, HttpSink responseBodySink) throws IOException {
+            responseBodySink.sink().close();
+
+            requestBodySourceRef.set(requestBodySource);
+            return serverHeadersListener;
+          }
+        }));
+    enableProtocol(Protocol.HTTP_2);
+
+    Call call = client.newCall(new Request.Builder()
+        .url(server.url("/"))
+        .duplex("POST")
+        .build());
+
+    Response response = call.execute();
+    HttpSink httpSink = response.httpSink();
+    httpSink.sink().writeUtf8("hey\n");
+    httpSink.headers(Headers.of("a", "android"));
+    httpSink.sink().writeUtf8("whats going on\n");
+    httpSink.headers(Headers.of("b", "blackberry"));
+    httpSink.sink().close();
+
+    // check what the server received
```

Comment:
I wanted to test things on the test thread so moved it here but I'm not using the latchParty with this one. Not sure what's best